qcodes/dataset/database_copy_paste.py

```diff
@@ -0,0 +1,397 @@
+
+
+def _copy_runs_table_entries(source_conn: SomeConnection,
+                             target_conn: SomeConnection,
+                             source_run_id: int,
+                             target_exp_id: int) -> int:
+    """
+    Copy an entire runs table row from one DB and paste it all
+    (expect the primary key) into another DB. The two DBs may not be the same.
+    Note that this function does not create a new results table
+
+    This function should be executed with an atomically guarded target_conn
+    as a part of a larger atomic transaction
+    """
+    runs_row_query = """
+                    SELECT *
+                    FROM runs
+                    WHERE run_id = ?
+                    """
+    cursor = source_conn.cursor()
+    cursor.execute(runs_row_query, (source_run_id,))
+    source_runs_row = cursor.fetchall()[0]
+    source_colnames = set(source_runs_row.keys())
+
+    # There might not be any runs in the target DB, hence we ask PRAGMA
+    cursor = target_conn.cursor()
+    cursor.execute("PRAGMA table_info(runs)", ())
+    tab_info = cursor.fetchall()
+    target_colnames = set([r['name'] for r in tab_info])
+
+    for colname in source_colnames.difference(target_colnames):
+        insert_column(target_conn, 'runs', colname)
+
+    # the first entry is "run_id"
+    sql_colnames = str(tuple(source_runs_row.keys()[1:])).replace("'", "")
+    sql_placeholders = sql_placeholder_string(len(source_runs_row.keys())-1)
+
+    sql_insert_values = f"""
+                         INSERT INTO runs
+                         {sql_colnames}
+                         VALUES
+                         {sql_placeholders}
+                         """
+    # the first two entries in source_runs_row are run_id and exp_id
+    values = tuple([target_exp_id] + [val for val in source_runs_row[2:]])
+
+    cursor = target_conn.cursor()
+    cursor.execute(sql_insert_values, values)
+
+    return cursor.lastrowid
+
+
+def _update_run_counter(target_conn: SomeConnection, target_exp_id) -> None:
+    """
+    Update the run_counter in the target DB experiments table
+    """
+    update_sql = """
+                 UPDATE experiments
+                 SET run_counter = run_counter + 1
+                 WHERE exp_id = ?
+                 """
+    cursor = target_conn.cursor()
+    cursor.execute(update_sql, (target_exp_id,))
+
+
+def _copy_layouts_and_dependencies(source_conn: SomeConnection,
+                                   target_conn: SomeConnection,
+                                   source_run_id: int) -> None:
+    """
+    Copy over the layouts and dependencies tables. Note that the layout_ids
+    are not preserved in the target DB, but of course their relationships are
+    (e.g. layout_id 10 that depends on layout_id 9 might be inserted as
+    layout_id 2 that depends on layout_id 1)
+    """
+    layout_query = """
+                   SELECT layout_id, run_id, "parameter", label, unit, inferred_from
```

Comment:
The word "parameter" gets syntax highlighted in DB browser, so I feared it might mean something. It does not seem to be among the 136 official keywords, though. https://www.sqlite.org/lang_keywords.html
🤷‍♂️